### PR TITLE
Remove the pear repository dependency

### PR DESCRIPTION
**Pull request by kabel**
Replaces HTTP_Request2 with GuzzleHttp and SimpleCAS with phpCAS.

### CODE REVIEW

File: src/UNL/Officefinder/Auth.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        return phpCAS::checkAuthentication();
+    }
+
+    public function forceAuthentication()
+    {
+        return phpCAS::forceAuthentication();
+    }
+
+    public function logout($url = '')
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: Is there a reason for the parameter if it isn't going to be used?

File: src/UNL/Officefinder/Auth.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    public function __construct()
+    {
+        if (phpCAS::isInitialized()) {
+            return;
+        }
+
+        phpCAS::client(CAS_VERSION_2_0, 'login.unl.edu', 443, 'cas', false);
+        phpCAS::setPostAuthenticateCallback([$this, 'handleLoginTicket']);
+        phpCAS::setSingleSignoutCallback([$this, 'handleLogoutTicket']);
+        phpCAS::setCasServerCACert(GuzzleHttp\default_ca_bundle());
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: This is nifty

File: src/UNL/Officefinder/Auth.php (hunk 1 above)
Comment (kabel) [inline src/UNL/Officefinder/Auth.php]: I probably should remove the param, just a b/c habit. The CAS server does not use it and CAS Protocol 3 says not to use it either.

File: src/UNL/Officefinder.php [hunk 3]
```diff
@@ -115,22 +114,7 @@ public function __construct($options = [])
...
-            ]);
-            $cacheDriver = new \Stash\Driver\FileSystem();
-            $cacheDriver->setOptions([
-                'path' => realpath(__DIR__ . '/../..') . '/tmp/simpleCAS_map',
-            ]);
-            $sessionMap = new SimpleCAS_SLOMap($cacheDriver);
-            $protocol->setSessionMap($sessionMap);
-
-            self::$auth = SimpleCAS::client($protocol);
+            self::$auth = new UNL_Officefinder_Auth();
```
Comment (mfairchild365) [inline src/UNL/Officefinder.php]: I'm guessing that making this static is to allow configuring the authentication mechanism. This lends itself to an interface design which is currently not in this PR.

File: src/UNL/Officefinder/Auth.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+        if (!$sessionId) {
+            $sessionId = $cache->getSlow($key);
+        }
+
+        if (!$sessionId) {
+            return false;
+        }
+
+        $this->removeSessionFromCache($ticket);
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: `getSessionFromCache()` implies that you will only get the session from cache, not remove it too. Perhaps this method should only do one thing, and removing the session from cache can be done in `handleLogoutTicket()`

File: src/UNL/Officefinder/Auth.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
+<?php
+
+class UNL_Officefinder_Auth
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: A general comment regarding methods in this class. Is there a reason that method comments that describe the purpose, parameters and return values were not included?

File: src/UNL/Officefinder/Auth.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        $key = $this->getCacheKeyFromTicket($ticket);
+        $value = session_id();
+        $this->getCache()->set($key, $value);
+        return $this;
+    }
+
+    public function handleLogoutTicket($ticket)
+    {
+        $sessionId = $this->getSessionFromCache($ticket);
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: It looks like you are implementing your own session mapping. If I remember correctly, phpCAS includes single logout logic based on making your session cookie name the ticket name. Out of curiosity, is there a reason you chose to roll your own here?

File: src/UNL/Peoplefinder/Cache.php [hunk 7]
```diff
@@ -59,4 +59,10 @@ public function set($key, $value, $expires = false)
 		$this->fastCache->set($key, $value, $expires);
 		$this->slowCache->save($value, $key);
 	}
+
+	public function remove($key)
```
Comment (mfairchild365) [inline src/UNL/Peoplefinder/Cache.php]: A method comment would be helpful

File: src/UNL/Peoplefinder/Record/Avatar.php [hunk 8]
```diff
@@ -155,17 +161,26 @@ protected function generatePersonUrl($options)
...
-            'adapter' => 'HTTP_Request2_Adapter_Curl',
-            'follow_redirects' => true,
-            'strict_redirects' => true,
+        $effectiveUrl = $profileIconUrl;
+        $onRedirect = function(
+            RequestInterface $request,
+            ResponseInterface $response,
+            UriInterface $uri
+        ) use (&$effectiveUrl) {
+            $effectiveUrl = (string) $uri;
```
Comment (mfairchild365) [inline src/UNL/Peoplefinder/Record/Avatar.php]: This is clever, but I wonder if `$effectiveUrl = $response->getEffectiveUrl()` would be easier to read

File: src/UNL/Peoplefinder/Record/Avatar.php [hunk 9]
```diff
@@ -181,6 +196,11 @@ protected function generatePersonUrl($options)
             $fallbackUrl = $effectiveUrl;
         }
 
+        // Do we have something Gravatar can use?
+        if (!$this->record->mail || !$this->record->eduPersonPrincipalName) {
```
Comment (mfairchild365) [inline src/UNL/Peoplefinder/Record/Avatar.php]: Nice addition

File: src/UNL/Officefinder/Auth.php (hunk 6 above)
Comment (kabel) [inline src/UNL/Officefinder/Auth.php]: I'm glad you ask. phpCAS uses the ticket from the CAS server as the new session id locally. This means that if someone learns your CAS ticket (which is trivial because it's in the querystring), they can hijack your session. By using our own generated session ids, we are doing a better job at keeping the login token secure and helping to ensure that the worst thing that can happen with the ticket is a forged logout request.

File: src/UNL/Officefinder/Auth.php (hunk 2 above)
Comment (kabel) [inline src/UNL/Officefinder/Auth.php]: Yep. I stumbled upon that when looking at the latest source.

File: src/UNL/Officefinder.php (hunk 3 above)
Comment (kabel) [inline src/UNL/Officefinder.php]: It's static just because it's already static. It wasn't worth fixing all the references to it. Static is not to allow configuration, but to better support the singleton pattern.

File: src/UNL/Officefinder/Auth.php (hunk 4 above)
Comment (kabel) [inline src/UNL/Officefinder/Auth.php]: I can agree with that.

File: src/UNL/Peoplefinder/Record/Avatar.php (hunk 8 above)
Comment (kabel) [inline src/UNL/Peoplefinder/Record/Avatar.php]: Unfortunately, the response object does not have that API. The only way to get the URI is by examining the request. The client frequently modifies the initial request (by cloning) leaving no way to get the _final_ version of the request that was used with the response. Guzzle just doesn't want to make it easy ;)

**Comment by kabel** [comment]
Better now.

File: src/UNL/Officefinder.php (hunk 3 above)
Comment (mfairchild365) [inline src/UNL/Officefinder.php]: Very similar to what you are doing, but 6.1 has an `on_stats` callback with the effective url. http://docs.guzzlephp.org/en/latest/request-options.html#on-stats

File: src/UNL/Peoplefinder/Record/Avatar.php (hunk 8 above)
Comment (mfairchild365) [inline src/UNL/Peoplefinder/Record/Avatar.php]: Very similar to what you are doing, but 6.1 has an `on_stats` callback with the effective url. http://docs.guzzlephp.org/en/latest/request-options.html#on-stats

File: src/UNL/Officefinder.php (hunk 3 above)
Comment (mfairchild365) [inline src/UNL/Officefinder.php]: Hmm. Perhaps a better pattern for a singleton would be something like `UNL_Officefinder_Auth::getInstance()` rather than putting construction logic in another class, but both of these are out of scope of the current PR.

File: src/UNL/Officefinder/Auth.php [hunk 10]
```diff
@@ -18,41 +21,68 @@ public function __construct()
...
+     * Handle a request that may match a backchannel SLO request from the CAS server
+     */
     public function handleSingleLogOut()
     {
         phpCAS::handleLogoutRequests(false);
     }
 
+    /**
+     * Checks if CAS authentication was previously done or has an active ticket
+     * @return boolean [description]
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: [description]?

File: src/UNL/Officefinder/Auth.php [hunk 11]
```diff
@@ -62,6 +92,11 @@ protected function getCache()
         return $this->cache;
     }
 
+    /**
+     * A callback for when the CAS library receives a valid service ticket
+     * @param  string $ticket [description]
```
Comment (mfairchild365) [inline src/UNL/Officefinder/Auth.php]: [description]?
